File: 프로그래머스/0/181918. 배열 만들기 4/배열 만들기 4.py
def solution(arr):
    stk = []
    i = 0
    
    # 파이썬은 for문 i값 조작 불가능
    while i < len(arr):
        if not stk:
            stk.append(arr[i])
            i += 1
        elif stk[-1] < arr[i]:
            stk.append(arr[i])
            i += 1
        elif stk[-1] >= arr[i]:
            stk.pop()
    
    return stk

File: 프로그래머스/0/181918. 배열 만들기 4/배열 만들기 4.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

// arr_len은 배열 arr의 길이입니다.
int* solution(int arr[], size_t arr_len) {
    // return 값은 malloc 등 동적 할당을 사용해주세요. 할당 길이는 상황에 맞게 변경해주세요.
    int* stk = (int*)malloc(sizeof(int) * arr_len);
    int stkidx = 0, arridx = 0;
    
    while(1){
        if(arridx == arr_len) break;
        
        if(stkidx == 0){
            stk[stkidx++] = arr[arridx++];
        }else if(stk[stkidx - 1] < arr[arridx]){
            stk[stkidx++] = arr[arridx++];
        }else if(stk[stkidx - 1] >= arr[arridx]){
            stk[stkidx--] = NULL;
        }
    }
    
    return stk;
}